### PR TITLE
use matrix to avoid copy my self in github action(pipeline as) code

### PR DESCRIPTION
**Pull request by SamYuan1990**
#### Type of change

- Improvement (improvement to code, performance, etc)

#### Description

- use matrix to avoid copy my self in github action(pipeline as) code

#### Additional details

n/A

#### Related issues

n/A

### CODE REVIEW

**Comment by SamYuan1990** [comment]
@denyeart , pre discussed on fabric contributor meeting, I opened this pr as a sample for metric usage. please help with pr review process.

I will squash this PR after a review, as I am not sure if split integration test into each condition as parameter, and the usage for make native is acceptable for maintainers or not.

**Comment by SamYuan1990** [comment]
The changes for github action can be found at https://github.com/hyperledger/fabric/actions/runs/2939710490

**Comment by SamYuan1990** [comment]
@denyeart , could you please ask some one take a look at this kind of idea? otherwise, I am going to close this pr.

**Comment by denyeart** [comment]
@SamYuan1990 Sorry missed this initially... I like the new approach.

Can you look into the native test failure though, both in this PR and at https://github.com/hyperledger/fabric/actions/runs/2939710490

**Comment by SamYuan1990** [comment]
> https://github.com/hyperledger/fabric/actions/runs/2939710490

ok, if it looks worth to discuss, I am going to rebase and update.

**Comment by SamYuan1990** [comment]
@denyeart , could you please help rerun azp check? and review this pr?

**Comment by denyeart** [comment]
Github Actions seems to handle many jobs well so I think it is fine to run the integration tests per directory now. It does make the pipeline simpler and more easily extensible, and the check output is more clear now since it shows the integration test directory.

What do you think @jcastrence

**Comment by jcastrence** [comment]
@denyeart At worst the matrix job can't take longer than the current workflow set up and it seems to be working fine so I don't see why we don't implement this strategy. I would however recommend using a more recent patch version of Go 1.18 and Ubuntu 22.04 as was suggested in [this PR](https://github.com/hyperledger/fabric-ca/pull/317).

File: .github/workflows/verify-build.yml [hunk 1]
```diff
@@ -43,9 +43,12 @@ jobs:
...
         name: Run Unit Tests
-  integration-tests-1:
-    name: Integration Tests 1
+  integration-tests:
+    name: Integration Tests
     needs: basic-checks
+    strategy:
+      matrix:
+        INTEGRATION_TEST_SUITE: [raft,pvtdata,ledger,lifecycle,e2e,gateway,idemix,pkcs11,configtx,configtxlator,sbe,nwo,msp,discovery,gossip,devmode,pluggable]
     runs-on: ubuntu-20.04
```
Comment (jcastrence) [inline .github/workflows/verify-build.yml]:
```suggestion
    runs-on: ubuntu-22.04
```

**Comment by jcastrence** [comment]
I'm confused as to why [this verify build run](https://github.com/hyperledger/fabric/actions/runs/3180262891) took 49 min, I thought that all the integration tests would be run in parallel?

**Comment by SamYuan1990** [comment]
> this verify build run

image

I suppose even if the jobs running in parallel, they still need to waiting for github agent ready to run the job.

**Comment by SamYuan1990** [comment]
As for sample: 
```
Requested labels: ubuntu-22.04
Job defined at: hyperledger/fabric/.github/workflows/verify-build.yml@refs/pull/3617/merge
Waiting for a runner to pick up this job...
```

**Comment by jcastrence** [comment]
> As for sample:
> 
> ```
> Requested labels: ubuntu-22.04
> Job defined at: hyperledger/fabric/.github/workflows/verify-build.yml@refs/pull/3617/merge
> Waiting for a runner to pick up this job...
> ```

This makes sense since GitHub Docs also says the number of concurrently running jobs is [dependent on runner availability](https://www.youtube.com/watch?v=J_6Tx13VPvU). With this in mind, I think it makes more sense to keep the faster integration tests grouped into the same jobs. We can still utilize the matrix strategy as it is definitely more readable and will be easier to maintain.

Alternatively, we can (and probably should) change the order of the integration test directory names as they appear in the strategy so that they are ordered slowest to fastest, since [the order of the variables determines job creation priority](https://docs.github.com/en/actions/using-jobs/using-a-matrix-for-your-jobs#defining-the-maximum-number-of-concurrent-jobs:~:text=The%20order%20of%20the%20variables%20in%20the%20matrix%20determines%20the%20order%20in%20which%20the%20jobs%20are%20created.%20The%20first%20variable%20you%20define%20will%20be%20the%20first%20job%20that%20is%20created%20in%20your%20workflow%20run.).

**Comment by SamYuan1990** [comment]
> > As for sample:
> > ```
> > Requested labels: ubuntu-22.04
> > Job defined at: hyperledger/fabric/.github/workflows/verify-build.yml@refs/pull/3617/merge
> > Waiting for a runner to pick up this job...
> > ```
> 
> This makes sense since GitHub Docs also says the number of concurrently running jobs is [dependent on runner availability](https://www.youtube.com/watch?v=J_6Tx13VPvU). With this in mind, I think it makes more sense to keep the faster integration tests grouped into the same jobs. We can still utilize the matrix strategy as it is definitely more readable and will be easier to maintain.
> 
> Alternatively, we can (and probably should) change the order of the integration test directory names as they appear in the strategy so that they are ordered slowest to fastest, since [the order of the variables determines job creation priority](https://docs.github.com/en/actions/using-jobs/using-a-matrix-for-your-jobs#defining-the-maximum-number-of-concurrent-jobs:~:text=The%20order%20of%20the%20variables%20in%20the%20matrix%20determines%20the%20order%20in%20which%20the%20jobs%20are%20created.%20The%20first%20variable%20you%20define%20will%20be%20the%20first%20job%20that%20is%20created%20in%20your%20workflow%20run.).

let me try.

File: .github/workflows/verify-build.yml [hunk 2]
```diff
@@ -43,115 +43,13 @@ jobs:
...
-        name: Install SoftHSM
-      - run: make integration-test INTEGRATION_TEST_SUITE="sbe nwo msp"
-        name: Run Integration Tests 7
-  integration-tests-8:
-    name: Integration Tests 8
-    needs: basic-checks
-    runs-on: ubuntu-20.04
+    strategy:
+      matrix:
+        INTEGRATION_TEST_SUITE: ["raft","pvtdata","ledger","lifecycle","e2e","gateway idemix pkcs11 configtx configtxlator","sbe nwo msp","discovery gossip devmode pluggable"]
```
Comment (jcastrence) [inline .github/workflows/verify-build.yml]:
```suggestion
        INTEGRATION_TEST_SUITE: ["raft","pvtdata","ledger","lifecycle","e2e","discovery gossip devmode pluggable","gateway idemix pkcs11 configtx configtxlator","sbe nwo msp"]
```

**Review by jcastrence — COMMENTED** [review]
I would change line 16 to `GO_VER: 1.18` and line 21 to `runs-on: ubuntu-22.04`

**Comment by SamYuan1990** [comment]
> I would change line 16 to `GO_VER: 1.18` and line 21 to `runs-on: ubuntu-22.04`

updated with environment value to control go version CI.

**Comment by denyeart** [comment]
Ok, based on the delay and large number of runners, and overhead per runner, I have to agree with Julian that it would be better to continue to group the integration test executions into a smaller set. But if it can be done with a matrix that would be an improvement over the existing approach.

**Comment by SamYuan1990** [comment]
> Ok, based on the delay and large number of runners, and overhead per runner, I have to agree with Julian that it would be better to continue to group the integration test executions into a smaller set. But if it can be done with a matrix that would be an improvement over the existing approach.

updated, so far keep the same group runner with previous.

**Review by denyeart — APPROVED** [review]

**Comment by denyeart** [comment]
Thanks @SamYuan1990 !

**Comment by denyeart** [comment]
@Mergifyio backport release-2.5

**Comment by denyeart** [comment]
@Mergifyio backport release-2.4

**Comment by denyeart** [comment]
@Mergifyio backport release-2.2

**Pull request merged**